lib/src/configuration.cc

```diff
@@ -127,105 +141,134 @@ HW::ParseResult Configuration::initialize(int argc, char *argv[],
-
+        config_file_ = lth_file::tilde_expand(HW::GetFlag<std::string>("config-file"));
         if (!config_file_.empty()) {
             parseConfigFile();
         }
-
-        if (enable_logging) {
-            setupLogging();
-        }
-
-        validateAndNormalizeConfiguration();
-        setAgentConfiguration();
-
-        initialized_ = true;
+    } else {
+        // No further processing or user interaction are required if
+        // the parsing outcome is HW::ParseResult::HELP or VERSION
     }
 
     return parse_result;
 }
 
-void Configuration::setStartFunction(
-        std::function<int(std::vector<std::string>)> start_function) {
-    start_function_ = start_function;
-}
+static void validateLogDirPath(const std::string& logfile) {
+    auto logdir_path = fs::path(logfile).parent_path();
 
-void Configuration::validateAndNormalizeConfiguration() {
-    // determine which of your values must be initalised
-    if (HW::GetFlag<std::string>("broker-ws-uri").empty()) {
-        throw Configuration::UnconfiguredError{ "broker-ws-uri value must be defined" };
-    } else if (HW::GetFlag<std::string>("broker-ws-uri").find("wss://") != 0) {
-        throw Configuration::UnconfiguredError { "broker-ws-uri value must start with wss://" };
+    if (fs::exists(logdir_path)) {
+        if (!fs::is_directory(logdir_path)) {
+            throw Configuration::Error { "log directory is not a directory" };
+        }
+    } else {
+        throw Configuration::Error {
+            (boost::format("cannot write to '%1%'; its parent directory does "
+                           "not exist") % logfile).str() };
     }
+}
 
-    if (HW::GetFlag<std::string>("ssl-ca-cert").empty()) {
-        throw Configuration::UnconfiguredError { "ssl-ca-cert value must be defined" };
-    } else if (!lth_file::file_readable(HW::GetFlag<std::string>("ssl-ca-cert"))) {
-        throw Configuration::UnconfiguredError { "ssl-ca-cert file not found" };
-    }
+void Configuration::setupLogging() {
+    logfile_ = HW::GetFlag<std::string>("logfile");
+    auto log_on_stdout = (logfile_ == "-");
+    auto loglevel = HW::GetFlag<std::string>("loglevel");
+    std::ostream *stream = nullptr;
 
-    if (HW::GetFlag<std::string>("ssl-cert").empty()) {
-        throw Configuration::UnconfiguredError { "ssl-cert value must be defined" };
-    } else if (!lth_file::file_readable(HW::GetFlag<std::string>("ssl-cert"))) {
-        throw Configuration::UnconfiguredError { "ssl-cert file not found" };
+    if (!log_on_stdout) {
+        // We should log on file
+        logfile_ = lth_file::tilde_expand(logfile_);
+
+        // NOTE(ale): we must validate the logifle path since we set
+        // up logging before calling validateAndNormalizeConfiguration
+        validateLogDirPath(logfile_);
+
+        logfile_fstream_.open(logfile_.c_str(), std::ios_base::app);
+        stream = &logfile_fstream_;
+    } else {
+        // Log on stdout by default
```

Comment:
We can drop this comment. We're not really doing anything here by default, we're logging to stdout because it was configured that way.